pkg/cmd/cli/cmd/loginoptions.go

```diff
@@ -0,0 +1,325 @@
+	if err := o.GatherProjectInfo(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Makes sure it has all the needed information about the server we are connecting to,
+// particularly the host address and certificate information. For every information not
+// present ask for interactive user input. Will also ping the server to make sure we can
+// connect to it, and if any problem is found (e.g. certificate issues), ask the user about
+// connecting insecurely.
+func (o *LoginOptions) GatherServerInfo() error {
+	// we need to have a server to talk to
+
+	if util.IsTerminal(o.Reader) {
+		for !o.serverProvided() {
+			defaultServer := defaultClusterURL
+			promptMsg := fmt.Sprintf("Please provide the server URL or just <enter> to use '%v': ", defaultServer)
+
+			server := util.PromptForStringWithDefault(o.Reader, defaultServer, promptMsg)
+			kclientcmd.DefaultCluster = clientcmdapi.Cluster{Server: server}
+		}
+	}
+
+	// we know the server we are expected to use
+
+	clientCfg, err := o.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	// ping to check if server is reachable
+
+	osClient, err := client.New(clientCfg)
+	if err != nil {
+		return osclientcmd.WrapClientErrors(err)
+	}
+
+	result := osClient.Get().AbsPath("/osapi").Do()
+	if result.Error() != nil {
+		err = osclientcmd.WrapClientErrors(result.Error())
+
+		// certificate issue, prompt user for insecure connection
+
+		if isServerCertificateSignedByUnknownAuthority(err) {
+			fmt.Println("The server uses a certificate signed by unknown authority. You can bypass the certificate check but it will make all connections insecure.")
+
+			clientCfg.Insecure = util.PromptForBool(os.Stdin, "Use insecure connections [y/N]? ")
+			if !clientCfg.Insecure {
+				return err
+			}
+		}
+	}
+
+	// we have all info we need, now we can have a proper Config
+
+	o.Config = clientCfg
+
+	o.gatherServerInfo = true
+	return nil
+}
+
+// Negotiate a bearer token with the auth server, or try to reuse one based on the
+// information already present. In case of any missing information, ask for user input
+// (usually username and password, interactive depending on the Reader).
+func (o *LoginOptions) GatherAuthInfo() error {
+	o.assertGatheredServerInfo()
+
+	if me, err := whoami(o.Config); err == nil && (!o.usernameProvided() || (o.usernameProvided() && o.Username == usernameFromUser(me))) {
+		o.Username = usernameFromUser(me)
+		fmt.Printf("Already logged into '%v' as '%v'.\n", o.Config.Host, o.Username)
+
+	} else {
+		// if not, we need to log in again
+
+		o.Config.BearerToken = ""
+		_, err := tokencmd.RequestToken(o.Config, o.Reader, o.Username, o.Password)
+		if err != nil {
+			return err
+		}
+
+		me, err := whoami(o.Config)
+		if err != nil {
+			return err
+		}
+		o.Username = usernameFromUser(me)
+		fmt.Printf("Logged into '%v' as '%v'.\n", o.Config.Host, o.Username)
+	}
+
+	// if trying to log in an user that's not the currently logged in, try to reuse an existing token
```

Comment:
add a TODO to inspect this and decide about safety and intent.  I definitely like trying existing tokens and I think we limit them to contexts that already relate the token and the server, it's a safe operation.